logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/EphemeralMessageDeletionHandler.kt

```diff
@@ -0,0 +1,158 @@
+package com.wire.kalium.logic.feature.message
+
+import com.wire.kalium.logic.data.id.ConversationId
+import com.wire.kalium.logic.data.message.Message
+import com.wire.kalium.logic.data.message.MessageRepository
+import com.wire.kalium.logic.functional.map
+import com.wire.kalium.logic.functional.onSuccess
+import com.wire.kalium.util.KaliumDispatcher
+import com.wire.kalium.util.KaliumDispatcherImpl
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.channels.produce
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.update
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import kotlinx.datetime.Clock
+import kotlin.coroutines.CoroutineContext
+
+
+interface EphemeralMessageDeletionHandler {
+    fun startSelfDeletion(conversationId: ConversationId, messageId: String)
+
+    fun observePendingMessageDeletionState(): Flow<Map<Pair<ConversationId, String>, SelfDeletionTimeLeft>>
+
+    fun enqueuePendingSelfDeletionMessages()
+}
+
+internal class EphemeralMessageDeletionHandlerImpl(
+    private val userSessionCoroutineScope: CoroutineScope,
+    private val kaliumDispatcher: KaliumDispatcher = KaliumDispatcherImpl,
+    private val messageRepository: MessageRepository
+) : EphemeralMessageDeletionHandler, CoroutineScope by userSessionCoroutineScope {
+    override val coroutineContext: CoroutineContext
+        get() = kaliumDispatcher.default
+
+    private val ephemeralMessageDeletionTimeLeftMapMutex = Mutex()
+    private val ephemeralMessageDeletionTimeLeftMap: MutableStateFlow<Map<Pair<ConversationId, String>, SelfDeletionTimeLeft>> =
+        MutableStateFlow(emptyMap())
+
+    override fun startSelfDeletion(conversationId: ConversationId, messageId: String) {
+        userSessionCoroutineScope.launch {
+            // in case client start deletion couple of times for the message, we secure it with mutex
+            // that will update the map with a corresponding key and default value, so that coroutine
+            // launched after the first winning the race simply returns
+            ephemeralMessageDeletionTimeLeftMapMutex.withLock {
+                val isSelfDeletionOutgoing = ephemeralMessageDeletionTimeLeftMap.value[conversationId to messageId] != null
+                if (isSelfDeletionOutgoing) return@launch
+
+                ephemeralMessageDeletionTimeLeftMap.update { currentState ->
+                    currentState + ((conversationId to messageId) to SelfDeletionTimeLeft(0))
+                }
+            }
+
+            messageRepository.getMessageById(conversationId, messageId).map { message ->
+                require(message is Message.Ephemeral)
```

Comment:
If some app tries to wrongly call a `SelfDeletingTimer` we could fail silently and just log it instead of crashing.